modules/core/arrow-test/src/main/kotlin/arrow/test/generators/Generators.kt

```diff
@@ -5,165 +5,173 @@ import arrow.core.*
+    Tuple10(genA.generate(), genB.generate(), genC.generate(), genD.generate(), genE.generate(), genF.generate(), genG.generate(), genH.generate(), genI.generate(), genJ.generate())
+}
 
 fun genIntPredicate(): Gen<(Int) -> Boolean> =
   Gen.int().let { gen ->
     /* If you ever see two zeros in a row please contact the maintainers for a pat in the back */
     val num = gen.generate().let { if (it == 0) gen.generate() else it }
     val absNum = Math.abs(num)
-    Gen.oneOf(listOf<(Int) -> Boolean>(
+    Gen.from(listOf<(Int) -> Boolean>(
       { it > num },
       { it <= num },
       { it % absNum == 0 },
       { it % absNum == absNum - 1 })
     )
   }
 
-fun <B> genOption(genB: Gen<B>): Gen<Option<B>> =
-  object : Gen<Option<B>> {
+fun <B> genOption(genB: Gen<B>): Gen<Option<B>> {
     val random = genIntSmall()
-    override fun generate(): Option<B> =
-      if (random.generate() % 20 == 0) None else Option.just(genB.generate())
-  }
+    return Gen.create {
+        if (random.generate() % 20 == 0) None else Option.just(genB.generate())
+    }
+}
 
-inline fun <reified E, reified A> genEither(genE: Gen<E>, genA: Gen<A>): Gen<Either<E, A>> =
-  object : Gen<Either<E, A>> {
-    override fun generate(): Either<E, A> =
-      Gen.oneOf(genE, genA).generate().let {
+inline fun <reified E: Any, reified A: Any> genEither(genE: Gen<E>, genA: Gen<A>): Gen<Either<E, A>> = Gen.create {
+    oneOf(genE, genA).random().map {
         when (it) {
-          is E -> Left(it)
-          is A -> Right(it)
-          else -> throw IllegalStateException("genEither incorrect value $it")
+            is E -> Left(it)
+            is A -> Right(it)
+            else -> throw IllegalStateException("genEither incorrect value $it")
         }
-      }
-  }
+    }.first()
+}
 
-inline fun <reified E, reified A> genValidated(genE: Gen<E>, genA: Gen<A>): Gen<Validated<E, A>> =
+inline fun <reified E : Any, reified A : Any> genValidated(genE: Gen<E>, genA: Gen<A>): Gen<Validated<E, A>> =
   Gen.create { Validated.fromEither(genEither(genE, genA).generate()) }
 
-inline fun <reified A> genTry(genA: Gen<A>, genThrowable: Gen<Throwable> = genThrowable()): Gen<Try<A>> = Gen.create {
+inline fun <reified A : Any> genTry(genA: Gen<A>, genThrowable: Gen<Throwable> = genThrowable()): Gen<Try<A>> = Gen.create {
   genEither(genThrowable, genA).generate().fold(
     { throwable -> Failure<A>(throwable) },
     { a -> Success(a) }
   )
 }
 
-fun <A> genNullable(genA: Gen<A>): Gen<A?> =
-  Gen.oneOf(genA, Gen.create { null })
+fun <A : Any> genNullable(genA: Gen<A>): Gen<A?> = object: Gen<A?> {
+    override fun always(): Iterable<A?> = emptyList()
+
+    override fun random(): Sequence<A?> = oneOf(genA, Gen.create { Option.empty<A>() }).random().map {
```

Comment:
We need to capture oneOf(..) as a field, and only call `random` here, to not reset the seed on every call.

Can't we use `Gen.create { null }` on 3.X?

Reply:
No, sequences cannot return null values.

Reply:
:( Can use use another marker value? Like, a literal value we create inside the class with an identity that will not overlap with A? Maybe compare using ===?

Reply:
I was wrong here also, sequence can have null values, but sequences generated throught `generateSequence` cannot (because null defines the end of the stream)